components/raftstore/src/store/util.rs

```diff
@@ -836,6 +837,153 @@ impl Display for MsgType<'_> {
     }
 }
 
+/// `RegionReadProgress` is used to keep track of the replica's `safe_ts`, the replica can handle a read
+/// request directly without requiring leader lease or read index iff `safe_ts` >= `read_ts` (the `read_ts`
+/// is usually stale i.e seconds ago).
+///
+/// `safe_ts` is updated by the `(apply index, safe ts)` item:
+/// ```
+/// if self.applied_index >= item.apply_index {
+///     self.safe_ts = max(self.safe_ts, item.safe_ts)
+/// }
+/// ```
+///
+/// For the leader, the `(apply index, safe ts)` item is publish by the `resolved-ts` worker periodically.
+/// For the followers, the item is sync periodically from the leader through the `CheckLeader` rpc and the
+/// raft heartbeat message
+///
+/// The intend is to make the item's `safe ts` larger (more up to date) and `apply index` smaller (require less data)
+#[derive(Debug)]
+pub struct RegionReadProgress {
+    // `core` used to keep track and update `safe_ts`, it should
+    // not be accessed outside to avoid dead lock
+    core: Mutex<RegionReadProgressCore>,
+    // The fast path to read `safe_ts` and `applied_index` without
+    // acquiring the mutex on `core`
+    safe_ts: AtomicU64,
+    applied_index: AtomicU64,
+}
+
+impl RegionReadProgress {
+    pub fn new(applied_index: u64, cap: usize) -> RegionReadProgress {
+        RegionReadProgress {
+            core: Mutex::new(RegionReadProgressCore::new(applied_index, cap)),
+            applied_index: AtomicU64::from(applied_index),
+            safe_ts: AtomicU64::from(0),
+        }
+    }
+
+    pub fn update_applied(&self, applied: u64) {
+        let mut core = self.core.lock().unwrap();
+        if let Some(ts) = core.update_applied(applied) {
+            self.safe_ts.fetch_max(ts, AtomicOrdering::SeqCst);
+        }
+        self.applied_index
+            .fetch_max(applied, AtomicOrdering::SeqCst);
+    }
+
+    pub fn update_safe_ts(&self, apply_index: u64, ts: u64) {
+        if apply_index == 0 || ts == 0 {
+            return;
+        }
+        let mut core = self.core.lock().unwrap();
+        if let Some(ts) = core.update_safe_ts(apply_index, ts) {
+            self.safe_ts.fetch_max(ts, AtomicOrdering::SeqCst);
+        }
+    }
+
+    pub fn applied_index(&self) -> u64 {
+        self.applied_index.load(AtomicOrdering::Acquire)
+    }
+
+    pub fn safe_ts(&self) -> u64 {
+        self.safe_ts.load(AtomicOrdering::Acquire)
+    }
+}
+
+#[derive(Debug)]
+struct RegionReadProgressCore {
+    applied_index: u64,
+    safe_ts: u64,
+    // `pending_items` is a *sorted* list of `(apply index, safe ts)` item
+    pending_items: VecDeque<(u64, u64)>,
```

Comment:
How long can the queue be? Is it possible not to store the queue at all?

Reply:
Currently, the capacity is `128`. 

I don't have a better idea, because the leader only sends the latest `(apply index, safe ts)` item, if the follower's apply index always lags behind the leader, without the queue the follower may never be able to update its safe ts. 

BTW, we only need one item instead of a queue if using `appiled ts` as `safe ts`, since the `appiled ts` can be updated by simply applying lagging logs.